### ISSUE TITLE
Layer outputs being computed multiple times

### ISSUE THREAD

**Issue by benanne** [Lasagne/Lasagne#104]
Lasagne `Layers` compute their outputs by calling `get_output()` recursively on their parents. When a network has a tree structure, which is usually the case, this works fine. But when a network has a directed acyclic graph (DAG) structure, i.e. some layers get their input from the same layer, this leads to `get_output()` being called multiple times on this layer. As a result the output expression of this shared input layer (and all layers further down) is constructed multiple times.

In terms of performance this is not an issue since all these computations are symbolic anyway. We count on Theano's smarts to notice that these sub-expressions are the same, so it only computes them once at execution time. This works great in most cases, but unfortunately there are exceptions:

* it fails when scan is used (see https://github.com/benanne/Lasagne/issues/17#issuecomment-70317085 and onwards, also https://github.com/benanne/Lasagne/issues/97#issuecomment-70718707). ~~This is because Theano does not seem to merge two scan operations that do the same computations.~~
EDIT: this is wrong, see @craffel's comment below. Theano does merge the scan operations, but it doesn't merge their *gradients*, which is also a problem.

* it fails when layer outputs are nondeterministic (because of e.g. dropout), see #97. Random variates are re-sampled for each call to `get_output()`, so Theano rightly does not merge the sub-expressions here. But of course this isn't what we want.

* there may be other cases that have not come up yet or have not been discovered yet.

If we are able to ensure that `get_output()` is called only once on each layer in a DAG, we can solve this issue. There are a couple of ways to do this:

* memoization: make `get_output_for()` remember the output associated with a given input after it's been computed for the first time. The second time it just returns the remembered output (which will be the same Theano expression). Doing this library-wide may result in some gotchas (people expecting code to be executed that will no longer be executed, etc.), so I don't know if this is a good idea. Making it a per-layer thing that should be enabled or disabled seems like it wouldn't solve the problem adequately (and is hard to implement as well), and also adds a cognitive burden for the user.

* making `get_output()` iterative instead of recursive: currently, `get_output()` calls `get_output()` on the parents and then feeds the result to `get_output_for()`. Instead of recursively computing output expressions like this, we could make this function iterative. We would still have to recurse to grab all the 'ancestors' of the layer (we can use `get_all_layers()` for this), but after that we can ensure that `get_output_for()` is only called once for each of them, even when the network structure is a DAG.

* there may be other ways to solve this, ideas are welcome!

I think I like the second option best so far because the interface would remain the same. I also feel like this might be a less intrusive change than introducing memoization. I have to admit I haven't really thought it through yet, though. We'll also have to think about how this will handle `InputLayers` and such (they don't define a `get_output_for()` method, only `get_output()`).

**Comment by f0k** [comment]
Gotcha: A network can also have multiple output layers. Your second solution wouldn't solve this because the user would have to call `get_output()` on the two output layers one after the other. Memoization would solve this, or pulling all the output propagation code out into a separate Model class (or yet another helper function in `lasagne/layers/helper.py`, or both -- we could have a Model class encapsulating all the helper functions in a sane manner just for convenience).

**Comment by benanne** [comment]
True, I didn't think about that.

**Comment by f0k** [comment]
Third solution (I don't like it, just wanted to throw it in): Give `get_output()` a `cache` argument that defaults to being initialized with an empty dictionary, is propagated along, and updated to store the output of each layer encountered. Because the dictionary is mutable, it can be used to propagate information both down and up the recursive calls. If provided by a user, she or he can ensure to use the same cache for each output layer's `get_output()` call in case of a multiple-output network. This solution feels hackish and requires users to actively take the necessary precautions when dealing with networks of multiple outputs (or even just when requiring the output of the code layer and the output layer of an autoencoder at once). I vote against it.

Fourth solution (briefly mentioned in my other commented): Create a function like:
```python
def get_output(layer_or_layers, input=None, **kwargs):
    """"
    Propagates data through a network and returns the output or outputs.
    @param layer_or_layers: The layer or list of layers to return output expressions for
    @param input: (as before)
    @return: An expression of list of expressions for the layer(s)'(s) output(s) given the input data
    """"
    # ...
```

`Layer.get_output()` should be removed then, otherwise the trap would still be open.

**Comment by craffel** [comment]
> This is because Theano does not seem to merge two scan operations that do the same computations.

Just for clarity here, I believe Theano will merge to scan operations (as long as there's no nondeterministic layer), but it doesn't seem to always merge scan gradients.

I think caching (option 1) is an OK but kind of messy idea.  Something like joblib.Memory would probably work.

I think I like the fourth solution the most because we could essentially write a graph traversal routine and then compute all the appropriate outputs and re-use them as necessary.  The function itself probably wouldn't be super short/clean/readable, but its usage would be.

**Comment by f0k** [comment]
> I think I like the fourth solution the most because we could essentially write a graph traversal routine and then compute all the appropriate outputs and re-use them as necessary. The function itself probably wouldn't be super short/clean/readable, but its usage would be.

Well, it would be:
```python
def get_output(layer_or_layers, input=None, **kwargs):
    all_layers = get_all_layers(layer_or_layers)
    if isinstance(input, dict):
        all_outputs = dict(input)
    elif input is not None:
        all_outputs = dict((layer, input) for layer in all_layers if isinstance(layer, InputLayer))
    else:
        all_outputs = dict((layer, layer.input_var) for layer in all_layers if isinstance(layer, InputLayer))
    for layer in all_layers[::-1]:
        if layer not in all_outputs:
            if isinstance(layer, MultipleInputsLayer):
                all_outputs[layer] = layer.get_output_for([all_outputs[input_layer] for input_layer in layer.input_layers], **kwargs)
            else:
                all_outputs[layer] = layer.get_output_for(all_outputs[layer.input_layer], **kwargs)
    if isinstance(layer_or_layers, list):
        return [all_outputs[layer] for layer in layer_or_layers]
    else:
        return all_outputs[layer_or_layyers]
```
It relies on the fact that `get_all_layers()` returns layers in reverse topological order. This should probably be documented and tested so people can rely on that.
Also it turns the polymorphism of `get_output()` into explicit type checks, which is not *that* nice... if we don't want that, we'd have to keep `Layer.get_output()` and code the helper function like this:
```python
def get_output(layer_or_layers, input=None, **kwargs):
    all_layers = get_all_layers(layer_or_layers)
    all_outputs = dict((layer, layer.get_output(input)) for layer in all_layers if isinstance(layer, InputLayer))
    for layer in all_layers[::-1]:
        if layer not in all_outputs:
            all_outputs[layer] = layer.get_output(all_outputs)
    if isinstance(layer_or_layers, list):
        return [all_outputs[layer] for layer in layer_or_layers]
    else:
        return all_outputs[layer_or_layyers]
```
This would be the least obstrusive solution as it would keep the current code as is. But then, as I said above, how do you know when it's safe to call `get_output()` on a layer and when to use the helper function instead?

**Comment by craffel** [comment]
> It relies on the fact that get_all_layers() returns layers in reverse topological order.

Ah, I didn't realize that, that potentially simplifies things assuming it works in all the cases we care about.

> Also it turns the polymorphism of get_output() into explicit type checks, which is not that nice...

Yeah, the check for regular layer vs. multiple inputs layer is kind of ugly.  But...

> how do you know when it's safe to call get_output() on a layer and when to use the helper function instead?

This makes me prefer the solution where layers no longer have a `get_output` method.  We could potentially make it so that all layers could accept a list of inputs or a single input - for `Layer`s which are not `MultipleInputLayers` the input list would be a singleton, right?

**Comment by f0k** [comment]
Actually, `get_all_layers()` also distinguishes between `Layer` and `MultipleInputsLayer` ([via ducktyping](https://github.com/benanne/Lasagne/blob/7cc670a617b8aea03c61c3faf71274c375194323/lasagne/layers/helper.py#L57-60)), so we don't lose anything if `get_output()` does the same. I'm not too fond of explicitly checking for `InputLayer` instances either, but I don't see a way around it.
What's ugly about this helper function business is that it couples some `Layer` internals (`input_layer`, `input_layers`, `input_var`) with the helper functions, but the only way around that would be pulling out all the layer connections into a separate `Lasagne` (err, `Model`) class that takes care of that, leaving the `Layer` just with what's needed for `get_output_for()` and `get_output_shape_for()`. That would be the cleanest solution, but not necessarily the most pragmatic. The other solution would be memoization, which would keep the current usage patterns unchanged, but be quite intrusive. What do the design goals tell us?

**Comment by dimatura** [comment]
@f0k That fourth solution looks pretty decent to me, better than memoization at least. I don't think assuming `input_layer` and `input_var` is that bad, specially if having these members is documented as a requirement in case someone wanted to write their own `InputLayer`.

**Comment by dnouri** [comment]
I think looking up input_layer, input_layers and input_var attributes is
fine.

Maybe, if we're not too fond of dealing with ducktyping or isinstance,
there's the option of writing down the interfaces we care about as
abstract base classes (as in Python's abc module).  So e.g. an
InputLayerInterface or InputLayerABC would be define that input layers
have input_var, and bang it's become super explicit.

Not sure memoization would work at all; because how would you deal with
deleting the memoized value again after you've used it -- say you want
to call get_output for a different input.  Have the user call some
global "forget" function?  That'd be fugly.

I too think that the fourth option is the nicest.  Layer.get_output()
should be deprecated, and it should just call the new get_output for a
while until it vanishes totally.

**Comment by benanne** [comment]
I agree that option 4 sounds like the best option.

This solves a considerable gotcha with a not-that-common-but-common-enough use case (multiple output layers), and having to write `get_output(layer)` instead of `layer.get_output()` isn't exactly a huge sacrifice in terms of usability. So I guess I can live with it if the latter disappears eventually.

At least we can tackle this problem now, while we don't have an official release yet. I guess we could make `Layer.get_output()` generate a warning for now?

Also, what about `Layer.get_output_shape()`? Obviously this pitfall does not apply to `get_output_shape()` (unless we want to support non-deterministic output shapes!), but maybe having that one as a method and `get_output()` as a function is a bit inconsistent.

**Comment by dimatura** [comment]
I think the multiple-output use case might become more popular in the future as neural nets are applied to structured output problems, so it is definitely worth supporting.

Regarding `get_output_shape`, keeping it together with `get_output` seems tidier.

**Comment by f0k** [comment]
> Not sure memoization would work at all; because how would you deal with
deleting the memoized value again after you've used it -- say you want
to call get_output for a different input.

The memoization would of course have to be dependent on all the arguments to `get_output()`. It would be something like a dictionary mapping `(input, kwargs)` to the output expression, but it would need both the keys and the values to be weak references so they disappear when the user gives up all references to either the input or output (we don't want to introduce a memory leak, and at that point we don't need to guarantee perfect reproducibility any more).

---

Anyway, I also think solution 4 is probably the best one (coincidentally, solution 4 already briefly came up in issue #4 :)). The recursive `get_output()` feels a lot more natural, but we're dealing with DAGs, not trees, so we need some way to deal with multiply-visited nodes.
For symmetry, we should move out `get_output_shape()` as well, and give it an optional `input_shape` argument that can be `None`, a single shape, or a dictionary of shapes. The only thing that bothers me is that usually, `get_output_shape()` doesn't need to traverse the graph at all, because each layer already knows its input shape when we merge #52. Maybe `Layer` should get an `output_shape` property that just calls `get_output_shape_for(self.input_shape)` (or even an attribute set at construction), and the external `get_output_shape()` helper function would just deal with propagating shapes that are different from the default shapes?

I've also thought some more about pulling out all the connections into an external class, and found that this would just create another kind of ugliness. I'll discuss that in #2.

**Comment by dnouri** [comment]
On 01/22/2015 12:35 PM, Jan Schlüter wrote:
>     Not sure memoization would work at all; because how would you deal with
>     deleting the memoized value again after you've used it -- say you want
>     to call get_output for a different input.
> 
> The memoization would of course have to be dependent on all the
> arguments to |get_output()|. It would be something like a dictionary
> mapping |(input, kwargs)| to the output expression, but it would need
> both the keys and the values to be weak references so they disappear
> when the user gives up all references to either the input or output (we
> don't want to introduce a memory leak, and at that point we don't need
> to guarantee perfect reproducibility any more).

Except that get_output() depends on the state of the network as well,
right?  So it'd have to include the layer's parameters in the cache key.
 Say I want to get the output for a validation set after each epoch:
without considering the network's state, I'd get the same output every time.

**Comment by f0k** [comment]
> Except that get_output() depends on the state of the network as well,
right?  So it'd have to include the layer's parameters in the cache key.
 Say I want to get the output for a validation set after each epoch:
without considering the network's state, I'd get the same output every time.

No, that dependency is symbolic anyway. The memoized output is just the output expression returned by `get_output()`, which will involve the same shared variables whether you call it after 10 epochs or 12 epochs of training. The only trouble we're trying to solve is that sometimes, Theano cannot identify equivalent output expressions, so we would like to return identical output expressions whenever we would have returned equivalent expressions (identical as in `id(a) == id(b)`, i.e., references to the same expression instance).

It's a different matter if you replace some network parameters by references to other variables in between two calls to `get_output()`. That's not exactly forbidden, so yes, we would have to include weak references to `self.get_params()` in the dictionary key as well.

Moving the graph traversal out of the `Layer` class still seems like a less brittle and easier understandable solution, though. As @benanne said, we haven't released anything yet, so we're free to change the API until it fits all use cases we hadn't thought through!

**Comment by dnouri** [comment]
On 01/23/2015 03:13 PM, Jan Schlüter wrote:
>     Except that get_output() depends on the state of the network as well,
>     right? So it'd have to include the layer's parameters in the cache key.
>     Say I want to get the output for a validation set after each epoch:
>     without considering the network's state, I'd get the same output
>     every time.
> 
> No, that dependency is symbolic anyway.

Oh, you're right.  I mixed it up, assumed `get_output` already evaluates
the expression, which it obviously does not.   I guess in that case
memoization makes sense again.

However, the `get_output` function that you proposed solves another
problem too, namely multiple output layers, so probably still preferable.

**Comment by benanne** [comment]
I think there is a consensus to move `get_output()` out of the `Layer` class. We should work on this once #52 is merged. The sooner we get this sorted out the easier it will be.

**Comment by benanne** [comment]
It's time! Anyone want to pick this up? If not I probably will, but not for at least another week or so.

**Comment by ebattenberg** [comment]
I like the caching approach.  Is this equivalent to a depth-first traversal starting at the output layer?  If so, do we plan to support multiple output layers using this methodology (basically just start the traversal from each output layer iteratively but share the same "cache" across iterations)?  I could see this being important for avoiding repeated work when using googlenet-style intermediate output taps.

**Comment by f0k** [comment]
> I like the caching approach. Is this equivalent to a depth-first traversal starting at the output layer?

Not really. The caching approach ("memoization" in the [opening comment](https://github.com/benanne/Lasagne/issues/104#issue-55041050)) would just remember any output expression ever created (as long as there are strong references to the input expressions) and reuse it if asked for an output expression given the same input expression(s). This would happen separately at each layer, and be remembered between calls to `get_output()`, so it would support the multiple-output case as well.

> If so, do we plan to support multiple output layers using this methodology

Yes, that was one of the main goals: https://github.com/benanne/Lasagne/issues/104#issuecomment-70873268

> (basically just start the traversal from each output layer iteratively but share the same "cache" across iterations)

Yes, that's what my proposed implementation does: https://github.com/benanne/Lasagne/issues/104#issuecomment-70891895
It uses the graph traversal code of `get_all_layers()`, which already supports starting from multiple output layers since a while, and it remembers and reuses the output expression per node visited for one `get_output(layer_or_layers)` call.

**Comment by ebattenberg** [comment]
@f0k,  Cool, thanks for catching me up.  I need to learn to read the entire thread.  

 I still don't quite understand the purpose of `MultipleInputsLayer`.    It takes a list of inputs and produces what?  Does it do concatenation and then produce a single output?  Or does it produce a list of corresponding outputs?

**Comment by benanne** [comment]
`MultipleInputsLayer` is the base class to use for layers that have multiple input layers, such as `ConcateLayer` and `ElemwiseSumLayer`. Its name is a little confusing in retrospect because it's reminiscent of `InputLayer`.

**Comment by ebattenberg** [comment]
@benanne  Thanks.  I just looked for it under input.py, didn't find it, and assumed it was a new layer type that @f0k was implementing elsewhere.  Should've grepped.

**Comment by avdnoord** [comment]
I must say I like the memoization approach a lot more than changing get_output. It's solves all the problems mentioned and wouldn't require users to change their code. All the changes are internally.

Also, if we change get_output, when a user calls get_output multiple times in their code the problem would still occur .

**Comment by benanne** [comment]
@avdnoord raises a great point, if you have your output layer (or multiple) and call `get_output()` on it multiple times with the same input, you will still get different expressions if there is any nondeterminism in the network (e.g. dropout). This is still a bit of a gotcha imo, that only memoization would solve (our iterative reimplementation of `get_output()` as a function would not). 

There don't really seem to be any use cases for having multiple expressions using the same input variable with different 'randomness', so there is no point to this behavior (all it can really do is confuse people). I think memoization would be safe to do (unless I missed something). It's definitely the solution that requires the least amount of changes. But maybe there are other pitfalls we haven't thought of? 

I think the two main solutions we're considering right now are
* memoization
* replace `Layer.get_output()` with `nn.layers.get_output()`, implemented iteratively

Both seem to have significant advantages and disadvantages, so we should try to reach a consensus of what we're going for. I think most people are currently in favour of the latter, but it clearly isn't a flawless solution either (and requires an API change).

**Comment by ebattenberg** [comment]
> I think the two main solutions we're considering right now are
> 
> memoization
> replace Layer.get_output() with nn.layers.get_output(), implemented iteratively

The memoization approach caches the symbolic output for a given input locally within the layer instance?  And the second approach is the one that @f0k proposed in code [above](https://github.com/benanne/Lasagne/issues/104#issuecomment-70891895)?  That version does a form of non-persistent memoization/caching that only exists during the call to get_output() function?

Can someone list the major pros and cons of each?  This thread is getting long. :)

**Comment by f0k** [comment]
> I think memoization would be safe to do (unless I missed something).

But it would be a little tricky. We want the memoization to forget entries once nobody holds a reference to the input (because then nobody has a chance to ask for the output anyway, and because we do not want to create a memory leak including shared variables and whatnot).
This would be feasible with a `weakref.WeakKeyDictionary`, but usually the associated output expression will contain a reference to the input, so we mustn't hold a strong reference to the output expression either. This should work if we use a `WeakKeyDictionary` with the values (the output expressions) wrapped in `weakref()`s, or if we invent a `WeakKeyWeakValueDictionary` that forgets things when either the key or the value are not covered by any strong reference. Either of this means we will also forget a memoized output expression if nobody holds a reference to the output expression. I'm not sure whether this would be good. At first glance sure, if nobody knows the output anyway, it doesn't matter if we can reproduce it. However, what happens if somebody keeps a part of the output expression, or keeps a compiled function that just has a reference to an optimized version of the output expression?

So unless proven otherwise, I'm now convinced that we would have to choose between an implementation that deliberately creates a memory leak, and an implementation that often remembers things across calls, but not always.

> Can someone list the major pros and cons of each?

Hmm, not me right now, but that would be useful if we want everybody to chime in. At least I've given some cons for memoization :)

**Comment by craffel** [comment]
I'm still on the side of the `lasagne.layers.get_output` option, primarily because I think it adds the least overhead to the library and is conceptually simplest from a development standpoint.  I don't think asking people to change their code from `l_out.get_output()` to `lasagne.layers.get_output(l_out)` is a lot to ask, particularly because we don't have a public release yet.  And, we could keep `l_out.get_output()` syntax temporarily with a deprecation warning.  Finally,

>  if you have your output layer (or multiple) and call get_output() on it multiple times with the same input, you will still get different expressions if there is any nondeterminism in the network (e.g. dropout).

This is actually the behavior I'd expect, and I think if it's well-documented this is not an issue.

All that being said, I am fine with either provided that whichever solution is chosen does the right thing.  :)

**Comment by avdnoord** [comment]
> > if you have your output layer (or multiple) and call get_output() on it multiple times with the same input, you will still get different expressions if there is any nondeterminism in the network (e.g. dropout).

> This is actually the behavior I'd expect, and I think if it's well-documented this is not an issue.

I don't know if this is the expected behavior. 
For example, say people want to compute extra kind of outputs to the training for monitoring the activation-std or sparsity in a certain layer. They will call get_output(l_out) and get_output(l_intermediate) and the intermediate output will be different for both expressions.

Or for example in the case multiple error-measures. For convenience, or for more readible code, people might call get_output multiple times in their code in different places/functions, etc.

> .. primarily because I think it adds the least overhead to the library and is conceptually simplest from a development standpoint.

Actually, can't we do the memoization in just a few lines in the base layer class? The user doesn't know that anything has changed, but now layer.get_ouput() actually works like they expect it to.

**Comment by f0k** [comment]
> This is actually the behavior I'd expect

Same for me.

> For convenience, or for more readible code, people might call get_output multiple times in their code in different places/functions

Okay, that's something I can imagine as well.

> Actually, can't we do the memoization in just a few lines in the base layer class? The user doesn't know that anything has changed, but now layer.get_ouput() actually works like they expect it to.

Yes, only the two base classes need to be changed, but show me the few lines of code! Given my previous thoughts, I guess they will introduce a problem. (This is not a rhetorical question, I think it would actually be helpful to have a draft to compare the implementations.)

> I'm still on the side of the `lasagne.layers.get_output` option, primarily because I think it adds the least overhead to the library and is conceptually simplest from a development standpoint.

That's my feeling as well. The `get_output` function I drafted above is easy to understand, while memoization will be difficult to follow as a developer and feel like black magic as a user. But as I said above, it would be great to actually have a draft of the memoization code in order to compare it.

**Comment by dnouri** [comment]
On 02/10/2015 09:19 PM, Jan Schlüter wrote:
>     I think memoization would be safe to do (unless I missed something).
> 
> But it would be a little tricky. We want the memoization to forget
> entries once nobody holds a reference to the input (because then nobody
> has a chance to ask for the output anyway, and because we do not want to
> create a memory leak including shared variables and whatnot).
> This would be feasible with a |weakref.WeakKeyDictionary|, but usually
> the associated output expression will contain a reference to the input,
> so we mustn't hold a strong reference to the output expression either.
> This should work if we use a |WeakKeyDictionary| with the values (the
> output expressions) wrapped in |weakref()|s, or if we invent a
> |WeakKeyWeakValueDictionary| that forgets things when either the key or
> the value are not covered by any strong reference. Either of this means
> we will also forget a memoized output expression if nobody holds a
> reference to the output expression. I'm not sure whether this would be
> good. At first glance sure, if nobody knows the output anyway, it
> doesn't matter if we can reproduce it.

Maybe a simpler implementation could stick the memoized output as an
attribute on the layer_input.  Then, when no one else holds a reference
to the input or output, these objects can go away.

I too think that it probably doesn't matter that a lost memoized output
can't be reproduced.  You start a new Python process and get a new
output doesn't seem to matter either.

> However, what happens if somebody
> keeps a part of the output expression, or keeps a compiled function that
> just has a reference to an optimized version of the output expression?

I guess the output would still stick around?  Are you counting this as a
memory leak?

> So unless proven otherwise, I'm now convinced that we would have to
> choose between an implementation that deliberately creates a memory
> leak, and an implementation that often remembers things across calls,
> but not always.

All in all, I'm not convinced we've got a big problem with memory here.
 Maybe it's because I underestimate:

- how much memory these Theano expressions require
- how often we expect to call get_output during one run/process

My understanding of the typical memory leak is code that will continue
to eat memory throughout your process run, maybe every time a particular
function is called.  I guess in our case it's more likely we'll have a
(small?) finite number of these inputs and outputs per run.  Or is there
anyone that's creating these things on the fly?  :-)

**Comment by f0k** [comment]
>>  However, what happens if somebody keeps a part of the output expression, or keeps a compiled function that just has a reference to an optimized version of the output expression?

> I guess the output would still stick around? Are you counting this as a memory leak?

No, on the contrary, I guess the reference to the original output expression as produced by the network would be dropped, and so would the memoization, so we couldn't reproduce the output expression any longer.

> I guess in our case it's more likely we'll have a (small?) finite number of these inputs and outputs per run.

The way I use it, yes. But what if somebody iterates over a long sequence of different input expressions (e.g., shared variables) fed into a network? That doesn't seem to be a clever way to do things, but still, if we want everything to be the same as it is now, just with the added benefit of reproducibility, we should discard each memoized input and output expression once it's not used any more. And if we agree on this point, I find it hard to come up with a solution that works.

**Comment by avdnoord** [comment]
> The way I use it, yes. But what if somebody iterates over a long sequence of different input expressions (e.g., shared variables) fed into a network?

I guess we could say that they are probably misusing Lasange. 
If there is a good use-case, we could add an argument "memoize" to the layer that is True by default? In the documentation we can add a line that says that output expressions are cached because of reproducibility+speed and that this could be turned for special use-cases?

**Comment by f0k** [comment]
> If there is a good use-case, we could add an argument "memoize" to the layer that is True by default?

What if you need the memoization for a single `get_output()` call (e.g., to avoid the non-merged graphs we saw in RNNs), but not across calls? Should there be a `forget()` method as well?

I don't know if there is a use case that would be hampered by a "non-forgetting" memoization, it just feels dangerous from a software design perspective. An explicit, external `get_output()` function cannot get back at us later. And memoization with a `forget()` method would be awkward because it would expose on one end what we're trying to hide on another.

**Comment by ebattenberg** [comment]
> I'm still on the side of the lasagne.layers.get_output option, primarily because I think it adds the least overhead to the library and is conceptually simplest from a development standpoint.

I agree with this sentiment.  I think doing anything too fancy here will defeat the 'simplicity' goal of Lasagne.  Memoizing and caching everything in the layer itself almost feels like we're trying to rebuild something like what Theano is already doing with its symbolic expression graph.  It seems like it has the potential to create a lot of headaches down the road.  When the output cache only lasts as long as the call to `get_output` a lot of potential confusion goes away.

**Comment by craffel** [comment]
> (This is not a rhetorical question, I think it would actually be helpful to have a draft to compare the implementations.) 

:+1: to this - @avdnoord @f0k do you each (or someone else) want to code up `layers.get_output` and memoization in `Layer.get_output` respectively? I think this would help a lot to actually figure out which is better/works in all the cases we care about.

**Comment by dimatura** [comment]
This is a sort of low-level detail, but I don't think @f0k 's code will work as-is, because (unless I'm missing something) you can't reverse a topologically sorted list and expect to get the topological sort of the reversed graph.

**Comment by f0k** [comment]
> you can't reverse a topologically sorted list and expect to get the topological sort of the reversed graph

Hmm. `get_all_layers()` does a breadth-first search starting from the output layers you give it. It will return a reversed topological ordering (for every directed edge *uv*, *v* will come before *u*). If I reverse it, I get the ordering I need (for every directed edge *uv*, *u* will come before *v*). Do you have a counterexample?

**Comment by dimatura** [comment]
l_in = lasagne.layers.InputLayer((16, 1, 20, 20), name='l_in')
l_conv1 = lasagne.layers.Conv2DLayer(l_in, 64, (5,5), border_mode='same', name='l_conv1')
l_drop1 = lasagne.layers.DropoutLayer(l_conv1, name='l_drop1')
I haven't slept much so I might not be making much sense. But I tried
```
l_conv2 = lasagne.layers.Conv2DLayer(l_drop1, 64, (5,5), border_mode='same', name='l_conv2')
l_drop2 = lasagne.layers.DropoutLayer(l_conv2, name='l_drop2')
concat = lasagne.layers.ConcatLayer((l_drop1, l_drop2), name='concat')
l_fc = lasagne.layers.DenseLayer(concat, 2, name='fc')
```
for which `get_all_layers()` gives the correct topological ordering
```
 ['fc', 'concat', 'l_drop1', 'l_drop2', 'l_conv1', 'l_conv2', 'l_in']
```
however when reversed this doesn't really do what we want, because you need to put `l_drop1` in `all_outputs` before processing `l_conv2`, which is what the code does now.

**Comment by f0k** [comment]
> I haven't slept much so I might not be making much sense.

Same disclaimer for me.

So we have:
```
l_in --> l_conv1 --> l_drop1 --> l_conv2 --> l_drop2 --> concat --> l_fc
                         \_________________________________/^
```

Now the problem is that `['fc', 'concat', 'l_drop1', 'l_drop2', 'l_conv1', 'l_conv2', 'l_in']` is **not** a correct topological ordering. If it was a reversed topological ordering (= a topological ordering of the reversed graph), then for every directed edge *uv*, *v* should come before *u*. This is not true for the edge `l_drop1 --> l_conv2`.

Good catch! While I think my reasoning was correct, `get_all_layers()` does not produce a reverse topological ordering to begin with. A correct solution would have been `['fc', 'concat', 'l_drop2', 'l_conv2', 'l_drop1', 'l_conv1', 'l_int']`, and reversing this would have been fine for `get_outputs`.

That's a bit unfortunate. I would actually propose to change `get_all_layers()` to produce a topological ordering, because that could be useful for a lot of things. Ideally, this would change the output of `get_all_layers()` for very special DAGs only (similar to @dimatura's clever construction), so it shouldn't break many existing models. As far as I see (but note my disclaimer above), stacks of layers and trees wouldn't need to change.

**Comment by dimatura** [comment]
Ok, I see what you're saying. One potential problem with changing the order of `get_all_layers` output is if someone has serialized their model with `get_all_parameters` under the old model and relies on that ordering being preserved. Not sure if anyone would really be affected.

For what it's worth I wasn't trying to be clever ;) This is actually a kind of architecture that is becoming common in computer vision to integrate information over different scales. So supporting it correctly is important :)

**Comment by f0k** [comment]
> One potential problem with changing the order of `get_all_layers` output is if someone has serialized their model with `get_all_parameters` under the old model and relies on that ordering being preserved.

Yes, that's why we should have been careful not to change anything for "usual" linear or tree-like architectures.

I say "should have been" because I now think there is no simple way to change `get_all_layers()` to produce a topological ordering. All variations of the current implementation can be fooled by the following:
```
l1 --> l2 --> l3 --> l4 --> l5
```
Then call `get_all_layers([l3, l5])`.
To solve this, one would need to grow partial orderings from every `layers_to_expand` and merge them when they meet.

So a simpler solution might be not to touch `get_all_layers()` at all, but to provide a separate `toposort()` function that takes a list of layers and sorts them.

> For what it's worth I wasn't trying to be clever ;)

Looked clever to me, because whether it fails or not depends on the order of entries in the `ConcatLayer` :)

**Comment by jsusskin** [comment]
This is a great discussion. Has there been any further progress teasing out which method works best?

**Comment by benanne** [comment]
We are most likely opting to turn `get_output()` into a top-level function (and the same will probably happen with all other functions that do graph traversal, i.e. do not operate on a single layer). How we'll achieve this is still a point of discussion, but we have a team meeting planned and this is of course one of the things we'll discuss. We hope to get it fixed soon!

**Comment by dimatura** [comment]
@jsusskin if you want a workaround right now, you can take the code by @f0k above but incorporate a proper topographical sorting to avoid the issue I described, works for me.

**Comment by jsusskin** [comment]
@dimatura and @f0k, that's a good idea. Will try this on Monday. One thing that we noticed was that we still observe this issue even if we take out the dropout layers we were using when we first noticed the overhead. So, it will be interesting to see if the proposed solution helps. Will report back during the week.
You guys are doing great work by the way.

**Comment by dimatura** [comment]
I'll try to find my code and put up a gist.

**Comment by f0k** [comment]
> I would actually propose to change `get_all_layers()` to produce a topological ordering, because that could be useful for a lot of things.

Another use case this would be useful for is Batch Normalization: After training, you need to determine the normalization factors (mean and std) for each batch normalization layer in a layer-wise fashion, starting from the input. It would be very practical to have `get_all_layers()` return the layers in topological order for this, if you're dealing with something more complex than a stack of layers.

The common solution to produce a topographic ordering is to do a depth-first search and prepend each node to the output list *once all its descendents have been processed*.

> Ideally, this would change the output of `get_all_layers()` for very special DAGs only [...], stacks of layers and trees wouldn't need to change.

Unfortunately, that algorithm would change the order of layers for trees, because the current algorithm is a breadth-first search.

> So a simpler solution might be not to touch `get_all_layers()` at all, but to provide a separate `toposort()` function that takes a list of layers and sorts them.

That would open up a can of worms. What's supposed to happen if I have a graph `l1 -> l2 -> l3` and then ask to `toposort([l1, l3])`?


A possible solution would be an additional `sorted` or `depth_first` argument to `get_all_params()` that defaults to `False`, but can be set to `True` to switch to a depth-first search that produces a topographic ordering (or a reverse topographic ordering). This wouldn't break any existing model serialization code, but would provide a convenient way to obtain a sorted list of layers.

**Comment by benanne** [comment]
I haven't written any code yet, but I've more or less come to the conclusion that implementing a `get_output()` that would work with a non-toposorted list of layers would effectively perform a toposort on it anyway. I think. I'm not sure.

So I think in the end making `get_all_layers` return a toposorted list of layers is probably the best option.

I also don't think we need a keyword argument for this. We have not released anything yet, so we shouldn't worry about backward compatibility too much. Having a keyword argument just so we can keep 'the old way' around seems like a bad idea. Yes, we will break people's code, but that's what happens when you use unreleased libraries :p

**Comment by f0k** [comment]
> Having a keyword argument just so we can keep 'the old way' around seems like a bad idea.

That's right. It's a small function, we can just pop it into our own code under a different name in order to be able to read our old models, we don't need to spoil our Lasagne with it.

**Comment by f0k** [comment]
Implementation suggestion (DFS avoiding the recursion):
```python
def get_all_layers_topo(layer):
    try:
        queue = deque(layer)
    except TypeError:
        queue = deque([layer])
    seen = set()
    done = set()
    result = []

    while queue:
        layer = queue[0]
        if layer in seen:
            queue.popleft()
            if layer not in done:
                result.append(layer)
                done.add(layer)
            continue
        seen.add(layer)
        if hasattr(layer, 'input_layers'):
            queue.extendleft(layer.input_layers)
        elif hasattr(layer, 'input_layer'):
            queue.appendleft(layer.input_layer)

    return result
```
Disclaimer: Haven't really tested it. It's just to see what it would look like.

**Comment by benanne** [comment]
seems to work fine (with `from collections import deque` added of course). But I guess it would be good to have some tests for this as well :) We'll need some 'exotic' architectures with branching and multiple output layers for this.

I guess the first step is now to replace `get_all_layers` by this implementation. Are you okay with doing the PR for that? That will be interface-breaking so I'll make sure to do a post on lasagne-users when that's merged.

Then the next thing is to turn `get_output()` into a function (as well as `get_output_shape()` I suppose). We can do that in a separate follow-up PR, I think.

**Comment by f0k** [comment]
> Are you okay with doing the PR for that?

Yes, sure. And yes, I'll add tests that hopefully cover all cases. Will do that on Monday, probably.

> That will be interface-breaking so I'll make sure to do a post on lasagne-users when that's merged.

I'd like to add a warning along with a way to disable it so everybody will know about the change and not blindly run into problems. This warning would be removed before the first release, of course.

> We can do that in a separate follow-up PR, I think.

Yes, that should be kept separate. That will touch a lot more code.

**Comment by benanne** [comment]
> I'd like to add a warning along with a way to disable it so everybody will know about the change and not blindly run into problems. This warning would be removed before the first release, of course.

Good call, I guess we should keep the old version around for a bit so peoples saved model parameter files aren't completely useless all of a sudden :)

**Comment by benanne** [comment]
@f0k are you planning to do the new `get_output()`? Otherwise I might take a stab at it, now that the new `get_all_layers()` is in place.

**Comment by jsusskin** [comment]
Here are several architectures with multiple output layers that I've developed in my own hack on top of lasagne that deals with this multiple outputs issue and seems to work fine. However, I'd be happy to stop using our hacks if this is dealt with.

1. Surrogate output layers at different depths of the net (to experiment with deep supervision)
2. Autoencoders with units in the encoder layer that need to predict labels (like Brian Cheung's work at Berkeley)
3. Different output layers to deal with multitask learning where the unit types and cost functions differ (so these may all be at the same depth but still need to be dealt with separately)
4. Various nets where an output layer serves as input to another part of the net (e.g. One type of output multiplicatively gates other units)

**Comment by f0k** [comment]
> @f0k are you planning to do the new `get_output()`?

Yes, I'll do that soon -- let's say this week.

@jsusskin: Thanks, that's a very convincing list!

**Comment by benanne** [comment]
And there was much rejoicing! yay!

**Comment by f0k** [comment]
Yes, it took some time from #97 to #182 :)

**Comment by craffel** [comment]
Lots of :+1:.   I will get #17 back on the table now.

**Comment by benanne** [comment]
Aww yiss
